chain/pruned_block_dispatcher.go

```diff
@@ -277,90 +281,149 @@ func (d *PrunedBlockDispatcher) connectToPeers() error {
 	if err != nil {
 		return err
 	}
-	peers, err = filterPeers(peers)
+	addrs, err := filterPeers(peers)
 	if err != nil {
 		return err
 	}
-	rand.Shuffle(len(peers), func(i, j int) {
-		peers[i], peers[j] = peers[j], peers[i]
+	rand.Shuffle(len(addrs), func(i, j int) {
+		addrs[i], addrs[j] = addrs[j], addrs[i]
 	})
 
-	// For each unbanned peer we don't already have a connection to, try to
-	// establish one, and if successful, notify the peer.
-	for _, peer := range peers {
-		d.peerMtx.Lock()
-		_, isBanned := d.bannedPeers[peer.Addr]
-		_, isConnected := d.currentPeers[peer.Addr]
-		d.peerMtx.Unlock()
-		if isBanned || isConnected {
-			continue
-		}
-
-		queryPeer, err := d.newQueryPeer(peer)
+	for _, addr := range addrs {
+		needMore, err := d.connectToPeer(addr)
 		if err != nil {
-			return fmt.Errorf("unable to configure query peer %v: "+
-				"%v", peer.Addr, err)
+			log.Debugf("Failed connecting to peer %v: %v", addr, err)
 		}
-		if err := d.connectToPeer(queryPeer); err != nil {
-			log.Debugf("Failed connecting to peer %v: %v",
-				peer.Addr, err)
-			continue
+		if !needMore {
+			return nil
 		}
+	}
 
-		select {
-		case d.peersConnected <- queryPeer:
-		case <-d.quit:
-			return errors.New("shutting down")
+	// We still need more addresses so we'll also invoke the
+	// `getnodeaddresses` RPC to receive random reachable addresses. We'll
+	// also filter out any that do not meet our requirements. The nil
+	// argument will return a default number of addresses, which is
+	// currently 8. We don't care how many addresses are returned as long as
+	// 1 is returned, since this will be polled regularly if needed.
+	nodeAddrs, err := d.cfg.GetNodeAddresses(nil)
+	if err != nil {
+		return err
+	}
+	addrs = filterNodeAddrs(nodeAddrs)
```

Comment:
Seems we're assuming here that all these connections will succeed rather than looping like we do above to continue to get new (unique) addresses until we reach out connection goal.

Reply:
I guess the assumption is that this will be run at regular intervals

Reply:
Yeah, `connectToPeers` is currently polled every 30s, so it would be handled there. The loop above for peer addresses works the same way.